### ISSUE TITLE
Allow multiple destinations for Session::consume

### ISSUE THREAD

**Issue by erenon** [Morgan-Stanley/binlog#4]
[Session::consume](https://github.com/Morgan-Stanley/binlog/blob/dev/include/binlog/Session.hpp#L126) takes an arbitrary object modeling the [mserialize::OutputStream](http://binlog.org/Mserialize.html#outputstream) concept, and writes metadata and data to it.

This decouples log rotation and destination selection from the Session. To send the data to multiple destinations, one simply needs to create a `MultiOutputStream`, which forwards calls to `write` to wrapped streams, e.g:

```
class MultiOutputStream {
public:
  MultiOutputStream(/* ... */);

  MultiOutputStream& write(const char* buf, std::streamsize size)
  {
    _file.write(buf, size);
    _monitor.write(buf, size);
    _console.write(buf, size);
    return *this;
  }

private:
  std::ostream _file;
  Monitoring _monitor;
  Console _console;
};
```

Writing a generic version of the above (e.g: `MultiOutputStream<Streams...>`) is non-trivial (e.g: how does the constructor look like, how to expose stats of children, etc), requires a lot of complexity, and probably cannot cover some use-cases. A specialized solution for ones use-case could be simpler and easier to use.

Therefore, unless a better design is suggested, create an example that shows how to create and use such a multi stream.

**Comment by loveyoupeng** [comment]
thanks for the reply, just wondering if the severity/category/writer check will be better to be abstracted as a dispatch strategy between session and output stream. 
My wild guess you try to leverage constexpr to push performance edge by static work trick, however, some form of configure files will be better to change behavior between different environment without change the code or generate different executables.

**Comment by erenon** [comment]
Do you mean selectively sending some events to some streams, while not sending others? e.g: sending everything to logfile, while also sending events above error severity (but only those) to console, while also sending events of category "monitoring" to the monitoring stream?

This can be done by a filter output stream, which forwards events allowed by a predicate (e.g: is severity above error, does this event belongs to the monitoring category, etc), while drops the others. Then the filters can be configured runtime. I'll open a ticket if that's what you are looking for.

**Comment by loveyoupeng** [comment]
definitely that's what I'm looking for, a typical use case will be logging "everything" to a local file while sending "ERROR" to alert agent, or logging all the request and response to a dedicated file while there is a leger for all logging events.

**Comment by erenon** [comment]
@loveyoupeng : This is fixed by #17 , please take a look to make sure this is what you meant.
